### PR TITLE
bpo-44392: Add Py_GenericAlias to C API docs

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Added docs for `Py_GenericAlias` and `Py_GenericAliasType`.
* ~Fixed the type for `Py_GenericAliasType` so the docs can compile.~ That's a separate PR now at GH-26739 which needs to be merged first.

https://bugs.python.org/issue44392

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
While writing the docs, I noticed that Py_GenericAliasType's return type in the stable ABI manifest file is wrong. It says it's a function but it should be a type/struct. The docs wouldn't compile until I fixed the type in stable_abi.dat.

**Comment by Fidget-Spinner** [comment]
BTW, I also noticed the stable ABI manifest exposes the wrong return type for Py_GenericAliasType. It exposes it as a function when it's a var/type/struct. So I fixed that in the PR too otherwise the docs weren't compiling. But that means the exported symbol for Windows is also modified.

File: Doc/c-api/typehints.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+:ref:`Union <types-union>`.  Only ``GenericAlias`` objects are currently exposed
+when including :file:`Python.h`.
+
+.. c:function:: PyObject* Py_GenericAlias(PyObject *origin, PyObject *args)
+
+   Return a :ref:`GenericAlias <types-genericalias>` object on success,
+   ``NULL`` on failure.  Equivalent to calling the Python class
+   :class:`types.GenericAlias`.  *origin* and *args* arguments set the
+   ``GenericAlias``\ 's ``__origin__`` and ``__args__`` attributes respectively.
+   *origin* should be a type, and *args* **must** be a :c:type:`PyTupleObject`.
```
Comment (JelleZijlstra) [inline Doc/c-api/typehints.rst]: I don't think that's true; the `setup_ga` method automatically packs anything that's not a tuple into a 1-tuple. Python itself passes a non-tuple here: https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Objects/abstract.c#L182

Comment (Fidget-Spinner) [reply Doc/c-api/typehints.rst]: Thanks for the catch and pointing this out. I should read the code more carefully next time.

**Comment by Fidget-Spinner** [comment]
> BTW, I also noticed the stable ABI manifest exposes the wrong return type for Py_GenericAliasType. It exposes it as a function when it's a var/type/struct. So I fixed that in the PR too otherwise the docs weren't compiling. But that means the exported symbol for Windows is also modified.

@encukou, is this alright with you and can I backport this? I was wondering if such a change to the stable ABI (for a bugfix) is okay.

**Comment by encukou** [comment]
Yes, fine with me, as long as it's backported to 3.10. This is a documentation change, except the function is not marked as DATA for the MSVC export. I'm not sure of the exact ABI impact of that, but it was clearly wrong before and we fixed things like this for 3.10.

cc @pablogsal: another ABI-related change for 3.10.

**Comment by Fidget-Spinner** [comment]
> Yes, fine with me, as long as it's backported to 3.10. This is a documentation change, except the function is not marked as DATA for the MSVC export. I'm not sure of the exact ABI impact of that, but it was clearly wrong before and we fixed things like this for 3.10.
> 
> cc @pablogsal: another ABI-related change for 3.10.

Great, thanks for your advice. I created a separate PR GH-26739 to fix the stable ABI errors and make backporting this docfix PR easier (it needs to land in 3.9 too). That will need to be merged first before this PR.

**Comment by pablogsal** [comment]
> cc @pablogsal: another ABI-related change for 3.10.

As we are before beta 4, this is still ok to me, but please, backport this before Thursday so it goes into beta 3.

File: Doc/c-api/typehints.rst [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+.. _typehintobjects:
+
+Objects for Type Hinting
+------------------------
+
+Various built-in types for type hinting are provided.  Currently,
+two types exist -- :ref:`GenericAlias <types-genericalias>` and
+:ref:`Union <types-union>`.  Only ``GenericAlias`` objects are currently exposed
+when including :file:`Python.h`.
```
Comment (gvanrossum) [inline Doc/c-api/typehints.rst]:
```suggestion
:ref:`Union <types-union>`.  Only ``GenericAlias`` is exposed to C.
```

File: Doc/c-api/typehints.rst [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+.. c:function:: PyObject* Py_GenericAlias(PyObject *origin, PyObject *args)
+
+   Return a :ref:`GenericAlias <types-genericalias>` object on success,
+   ``NULL`` on failure.  Equivalent to calling the Python class
```
Comment (gvanrossum) [inline Doc/c-api/typehints.rst]:
```suggestion
   Create a :ref:`GenericAlias <types-genericalias>` object.
   Equivalent to calling the Python class
```

File: Doc/c-api/typehints.rst [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+   :class:`types.GenericAlias`.  The *origin* and *args* arguments set the
+   ``GenericAlias``\ 's ``__origin__`` and ``__args__`` attributes respectively.
+   *origin* should be a :c:type:`PyTypeObject*`, and *args* can be a
+   :c:type:`PyTupleObject*` or any ``PyObject*``.  If *args* passed is
+   not a tuple, a 1-tuple is automatically constructed and ``__args__`` is set
+   to ``(args,)``.
+   Minimal checking is done for the arguments, so the function will succeed even
+   if *origin* is not a type.
+   The ``GenericAlias``\ 's ``__parameters__`` attribute is constructed lazily
+   from ``__args__``.
```
Comment (gvanrossum) [inline Doc/c-api/typehints.rst]:
```suggestion
   from ``__args__``.
   On failure, an exception is raised and ``NULL`` is returned.
```

File: Doc/c-api/typehints.rst [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+   Here's an example of how most types in the standard library use
+   ``Py_GenericAlias``.  This allows a type to be subscripted::
```
Comment (gvanrossum) [inline Doc/c-api/typehints.rst]:
```suggestion
   Here's an example of how to make an extension type generic::
```

File: Doc/c-api/typehints.rst [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+      static PyMethodDef my_obj_methods[] = {
+          {"__class_getitem__", (PyCFunction)Py_GenericAlias, METH_O|METH_CLASS, "See PEP 585"}
+      }
+
```
Comment (gvanrossum) [inline Doc/c-api/typehints.rst:37]:
```suggestion
          ...
      }
      static PyTypeObject MyObject_Type = {
          // The usual
          ...
          tp_methods = my_obj_methods,
          ...
      };
```

**Review by gvanrossum — COMMENTED** [review]
Thanks! Here are some suggestions. What do you think? Otherwise this looks good!

**Comment by Fidget-Spinner** [comment]
> Thanks! Here are some suggestions. What do you think? Otherwise this looks good!

Thanks for the review and suggestions. I think they make sense so they're added.

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
I don't know why the Address Sanitizer test is failing but I don't think it was your code, so I've merged regardless. Now on to the backport(s).

**Comment by Fidget-Spinner** [comment]
> I don't know why the Address Sanitizer test is failing but I don't think it was your code, so I've merged regardless. Now on to the backport(s).

Yeah I noticed it passed at first, then failed after I added a blurb entry, which probably means it was something else.

This doesn't affect much, but I initially split off the C API changes into GH-26739 which was supposed to be merged first (along with a separate news entry). However, since you already merged this PR, we don't need that anymore.

Thanks for the merge, and I'll work on the docs only backport to 3.9 :).

**Pull request merged**